[SPARK-19335][SPARK-38200][SQL] Add upserts for writing to JDBC

EnricoMi opened the pull request:
### What changes were proposed in this pull request?
This is a follow-up on #16685 and #16692.

Implements upsert mode for `SaveMode.Append` of the MySql, MsSql, and Postgres JDBC source.

See #41611 for an alternative using the `MERGE INTO` command (not supported by MySql).

### Why are the changes needed?
The JDBC writer only supports either truncating the existing table or inserting. Duplicates, i.e. rows with identical values in the primary or unique index columns, cause an exception, permitting updating existing and inserting new rows.

Re-evaluating a partition due to executor loss will insert rows that have been inserted in an earlier attempt, which kills the entier Spark job.

### Does this PR introduce _any_ user-facing change?
This adds `upsert` and `upsertKeyColumns` options for `SaveMode.Append` of the JDBC source.

### How was this patch tested?
Tests in `JdbcSuite` and integration suites.

EnricoMi committed:
Implement upsert for MySQL

EnricoMi committed:
Code cleanup

EnricoMi committed:
Implement upsert for MsSqlServer

EnricoMi committed:
Move upsert tests into trait

EnricoMi committed:
Fix non-existing upsert test for Postgres

EnricoMi committed:
Add upsert concurrency integration test

EnricoMi committed:
Add tests with varying column order

EnricoMi committed:
Add test with varying column order, sketch more tests

EnricoMi committed:
Revert empty line removal, fix scalastyle error

EnricoMi committed:
Refactor tableDoesNotSupportError to reuse in tableDoesNotSupportUpsertsError

EnricoMi committed:
Fix after merge master

EnricoMi committed:
Fix after merge master

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -336,6 +336,17 @@ abstract class JdbcDialect extends Serializable with Logging {
+      columns: Array[StructField],
+      isCaseSensitive: Boolean,
+      options: JDBCOptions): String =
+    throw new UnsupportedOperationException("upserts are not supported")
```
Could you use `SparkUnsupportedOperationException`.

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -132,9 +129,32 @@ object JdbcUtils extends Logging with SQLConfHelper {
   }
 
+  def getUpsertStatement(
+    table: String,
+    rddSchema: StructType,
+    tableSchema: Option[StructType],
+    isCaseSensitive: Boolean,
+    dialect: JdbcDialect,
+    options: JDBCOptions): String = {
```
fix indentation, see https://github.com/databricks/scala-style-guide?tab=readme-ov-file#spacing-and-indentation

MaxGekk commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/UpsertTests.scala:
```diff
@@ -0,0 +1,137 @@
+    assert(actual === expected)
+  }
+
+  test(s"Upsert concurrency") {
```
```suggestion
  test("Upsert concurrency") {
```

EnricoMi committed:
Fix after merge master

EnricoMi committed:
Apply code review comments
- Use SparkUnsupportedOperationException
- Remove unused string interpolation
- Fix indentation

LuciferYang commented:
cc @beliefer FYI

beliefer commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:
```diff
@@ -58,17 +59,20 @@ class JdbcRelationProvider extends CreatableRelationProvider
               truncateTable(conn, options)
               val tableSchema = JdbcUtils.getSchemaOption(conn, options)
-              saveTable(df, tableSchema, isCaseSensitive, options)
+              saveTable(df, tableSchema, isCaseSensitive, upsert, options)
```
It looks strange that apply `upsert` if `SaveMode` is `Overwrite` here.

beliefer commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:
```diff
@@ -58,17 +59,20 @@ class JdbcRelationProvider extends CreatableRelationProvider
               dropTable(conn, options.table, options)
               createTable(conn, options.table, df.schema, isCaseSensitive, options)
-              saveTable(df, Some(df.schema), isCaseSensitive, options)
+              saveTable(df, Some(df.schema), isCaseSensitive, upsert, options)
```
ditto

beliefer reviewed (commented):
I have no idea. It seems we should add a new SaveMode `Upsert`.
